pkg/reconciler/buildrun/resources/params.go

```diff
@@ -53,3 +65,511 @@ func overrideParams(originalParams []buildv1alpha1.ParamValue, overrideParams []
+					} else {
+						taskRunParam.Value.ArrayVal = append(taskRunParam.Value.ArrayVal, envVarExpression)
+					}
+				} else if value.SecretValue != nil {
+					envVarName, err := addSecretEnvVar(taskRun, paramValue.Name, value.SecretValue.Name, value.SecretValue.Key)
+					if err != nil {
+						return err
+					}
+
+					envVarExpression := fmt.Sprintf("$(%s)", envVarName)
+					if value.SecretValue.Format != nil {
+						taskRunParam.Value.ArrayVal = append(taskRunParam.Value.ArrayVal, strings.ReplaceAll(*value.SecretValue.Format, "${SECRET_VALUE}", envVarExpression))
+					} else {
+						taskRunParam.Value.ArrayVal = append(taskRunParam.Value.ArrayVal, envVarExpression)
+					}
+				} else if value.Value != nil {
+					taskRunParam.Value.ArrayVal = append(taskRunParam.Value.ArrayVal, *value.Value)
+				} else {
+					// this error should never happen because we validate this upfront in ValidateBuildRunParameters
+					return fmt.Errorf("unexpected parameter without any value: %s[%d]", parameterDefinition.Name, index)
+				}
+			}
+		} else if parameterDefinition.Defaults == nil {
+			// this error should never happen because we validate this upfront in ValidateBuildRunParameters
+			return fmt.Errorf("unexpected parameter without any value: %s", parameterDefinition.Name)
+		} else {
+			// we tolerate this for optional parameters, we enter this code path if a user provides a paramValue without any values. The theoretic but
+			// not documented use case is that a Build specifies values for an optional parameter (which has defaults in the strategy). Then one can
+			// set a paramValue without values in the BuildRun to get back to the default values.
+			return nil
+		}
+	}
+
+	taskRun.Spec.Params = append(taskRun.Spec.Params, taskRunParam)
+
+	return nil
+}
+
+func generateEnvVarName(prefix string) (string, error) {
+	result := prefix
+
+	for i := 0; i < 5; i++ {
+		num, err := rand.Int(rand.Reader, EnvVarNameSuffixCharsCount)
+		if err != nil {
+			return "", err
+		}
+		result += string(EnvVarNameSuffixChars[num.Int64()])
+	}
+
+	return result, nil
+}
+
+// addConfigMapEnvVar modifies all steps which are referencing a parameter name in their command, args, or environment variable values,
+// to contain a mapped environment variable for the ConfigMap key. It returns the name of the environment variable name.
+func addConfigMapEnvVar(taskRun *pipeline.TaskRun, paramName string, configMapName string, configMapKey string) (string, error) {
+	// check if an existing step references the same ConfigMap key
+	envVarName := ""
+
+stepLookupLoop:
+	for _, step := range taskRun.Spec.TaskSpec.Steps {
+		for _, env := range step.Env {
+			if strings.HasPrefix(env.Name, "SHP_CONFIGMAP_PARAM_") && env.ValueFrom != nil && env.ValueFrom.ConfigMapKeyRef != nil && env.ValueFrom.ConfigMapKeyRef.LocalObjectReference.Name == configMapName && env.ValueFrom.ConfigMapKeyRef.Key == configMapKey {
+				envVarName = env.Name
+				break stepLookupLoop
```

Comment:
first practical application of the use of labels I've seen outside an academic / golang instructional exercise :-)  

that said, in cross referencing this with the godoc for this method, I got a bit confused ... particularly, it mentions modifying **ALL** steps which reference a parameter

the break here suggests that only one step can have and env with the prefix SHP_CONFIGMAP_PARAM_ etc. ref

what am I missing?  assuming I am missing something, would a comment here to clarify vs. what is in the method's godoc be viable?

Reply:
Will extending the comments, this first loop just checks if any step already references an environment variable with the wanted ConfigMap key. This happens when a user uses the same key for more than one parameter value.

Reply:
Done.

Reply:
that made it clearer for me thanks